nir/src/main/scala/scala/scalanative/nir/Mangle.scala

```diff
@@ -127,6 +127,7 @@ object Mangle {
 
     def mangleIdent(id: String): Unit = {
       str(id.length)
+      if (id.head.isDigit || id.head == '$') str('$')
```

Comment:
So concretely, the problem was that the mangling for something like `id = "3foo"` was `43foo`, which was interpreted as a length of `43` and then not enough characters afterwards. Another way to say it is that the encoding was not injective.

Clearly the solution proposed here solves this issue in isolation, but I agree with @LeeTibbert's concerns over backward compat of the IR. I don't think we need to bump the minor IR version *once more*, since we have not released a public version (against which libraries would then be compiled and published) since the last bump. However, we should make sure that we can still read IR that was built by 0.4.0-M2.

`$` is too common a character; it is generated by the compiler for special characters, so an identifier like `+` is already mangled as `$plus`. We could however use another character that is never emitted by scalac under any circumstance. I suggest to use the character `-` for this purpose. Indeed, if an identifier contains a `-` in user-space, scalac encodes it as `$minus`, so at this point in the pipeline, `id` never contains a `-`. And `-` is a nice separation character: it would produce `4-3foo` for the example above.

If you try `-`, either it works and we're happy, or it crashes later because `-` would not be a valid ident for LLVM somewhere inside. In that case we'll have to think more.

Reply:
It works out well with `-` symbol instead of `$`. 
One thing I've noticed there was NIR file generated with name `DigitIdentifiersSuite$1Test$$minus1.nir` for ``` sealed trait `-1` {...}```. I'm not sure can it cause some problems in future.